File: leetcode/algorithms/350_intersection_of_two_arrays_2/main.js
/**
 * Complexities:
 *   N - The size of `nums1`
 *   M - The size of `nums2`
 *   - Time Complexity: O(N * M)
 *   - Space Complexity: O(N + M)
 */
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
var intersect = function (nums1, nums2) {
  const result = [];
  const numsMap = new Map();

  for (const num1 of nums1) {
    if (numsMap[num1]) {
      numsMap[num1] += 1;
    } else {
      numsMap[num1] = 1;
    }
  }

  for (const num2 of nums2) {
    if (numsMap[num2] > 0) {
      result.push(num2);
      numsMap[num2] -= 1;
    }
  }

  return result;
};


// Solution
/**
 * Solution 1
 *
 * Map
 *
 * Complexities:
 *   N - The size of `nums1`
 *   M - The size of `nums2`
 *   - Time Complexity: O(N + M)
 *   - Space Complexity: O(min(N, M))
 */
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
var solution1 = function (nums1, nums2) {
  if (nums1.length > nums2.length) {
    [nums1, nums2] = [nums2, nums1];
  }

  const count = new Map();
  for (const num of nums1) {
    count.set(num, (count.get(num) ?? 0) + 1);
  }

  const result = [];
  for (const num of nums2) {
    const remaining = count.get(num);
    if (remaining > 0) {
      result.push(num);
      count.set(num, remaining - 1);
    }
  }

  return result;
};

/**
 * Solution 2
 *
 * Sorting + Two Pointers
 *
 * Complexities:
 *   N - The size of `nums1`
 *   M - The size of `nums2`
 *   - Time Complexity: O(N * logᴺ + M * logᴹ)
 *   - Space Complexity: O(max(logᴺ, logᴹ))
 */
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
var solution2 = function (nums1, nums2) {
  nums1.sort((a, b) => a - b);
  nums2.sort((a, b) => a - b);

  const result = [];
  let i = 0,
    j = 0;

  while (i < nums1.length && j < nums2.length) {
    if (nums1[i] < nums2[j]) {
      i++;
    } else if (nums1[i] > nums2[j]) {
      j++;
    } else {
      result.push(nums1[i]);
      i++;
      j++;
    }
  }

  return result;
};

File: leetcode/algorithms/350_intersection_of_two_arrays_2/main.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class IntersectionOfTwoArraysII {
    /**
     * Complexities:
     *   N - The size of `nums1`
     *   M - The size of `nums2`
     *   - Time Complexity: O(N * M)
     *   - Space Complexity: O(N + M)
     */
    public int[] intersect(int[] nums1, int[] nums2) {
        List<Integer> result = new ArrayList<>();
        Map<Integer, Integer> numsMap = new HashMap<>();

        for (Integer num1 : nums1) {
            if (numsMap.get(num1) == null) {
                numsMap.put(num1, 1);
            } else {
                numsMap.put(num1, numsMap.get(num1) + 1);
            }
        }

        for (Integer num2 : nums2) {
            if (numsMap.get(num2) != null && numsMap.get(num2) > 0) {
                result.add(num2);
                numsMap.put(num2, numsMap.get(num2) - 1);
            }
        }

        return result.stream().mapToInt(Integer::intValue).toArray();
    }


    // Solution
    /**
     * Solution 1
     *
     * Map
     *
     * Complexities:
     *   N - The size of `nums1`
     *   M - The size of `nums2`
     *   - Time Complexity: O(N + M)
     *   - Space Complexity: O(min(N, M))
     */
    public int[] solution1(int[] nums1, int[] nums2) {
        if (nums1.length > nums2.length) {
            return intersect(nums2, nums1);
        }

        Map<Integer, Integer> count = new HashMap<>();
        for (int x : nums1) {
            count.merge(x, 1, Integer::sum);
        }

        int[] tmp = new int[nums1.length];
        int k = 0;
        for (int x : nums2) {
            int c = count.getOrDefault(x, 0);
            if (c > 0) {
                tmp[k++] = x;
                count.put(x, c - 1);
            }
        }

        return Arrays.copyOf(tmp, k);
    }

    /**
     * Solution 2
     *
     * Sorting + Two Pointers
     *
     * Complexities:
     *   N - The size of `nums1`
     *   M - The size of `nums2`
     *   - Time Complexity: O(N * logᴺ + M * logᴹ)
     *   - Space Complexity: O(1)
     */
    public int[] solution2(int[] nums1, int[] nums2) {
        Arrays.sort(nums1);
        Arrays.sort(nums2);

        int i = 0, j = 0, k = 0;
        int[] tmp = new int[Math.min(nums1.length, nums2.length)];

        while (i < nums1.length && j < nums2.length) {
            if (nums1[i] < nums2[j]) {
                i++;
            } else if (nums1[i] > nums2[j]) {
                j++;
            } else {
                tmp[k++] = nums1[i];
                i++;
                j++;
            }
        }

        return Arrays.copyOf(tmp, k);
    }

    /**
     * Solution 3
     *
     * Constraints
     *
     * Complexities:
     *   N - The size of `nums1`
     *   M - The size of `nums2`
     *   - Time Complexity: O(N + M)
     *   - Space Complexity: O(1)
     */
    public int[] solution3(int[] nums1, int[] nums2) {
        int[] count = new int[1001];
        for (int x : nums1) {
            count[x]++;
        }

        int[] tmp = new int[Math.min(nums1.length, nums2.length)];
        int k = 0;
        for (int x : nums2) {
            if (count[x] > 0) {
                tmp[k++] = x;
                count[x]--;
            }
        }

        return Arrays.copyOf(tmp, k);
    }
}
